src/main/java/com/microsoft/sqlserver/jdbc/SQLServerDriver.java

```diff
@@ -548,17 +548,45 @@ String getClassNameLogging() {
 
     private final static java.util.logging.Logger drLogger = java.util.logging.Logger
             .getLogger("com.microsoft.sqlserver.jdbc.internals.SQLServerDriver");
+    private static java.sql.Driver mssqlDriver = null;
     // Register with the DriverManager
     static {
         try {
-            java.sql.DriverManager.registerDriver(new SQLServerDriver());
+            register();
         } catch (SQLException e) {
             if (drLogger.isLoggable(Level.FINER) && Util.IsActivityTraceOn()) {
                 drLogger.finer("Error registering driver: " + e);
             }
         }
     }
 
+    /*
+     * Registers the driver with DriverManager. No-op if driver is already registered.
+     */
+    public static void register() throws SQLException {
+        if (!isRegistered()) {
+            mssqlDriver = new SQLServerDriver();
```

Comment:
I would say its better to not make this static but an instance method of the Driver so you can then use 
```
driver = new SQLServerDriver()
driver.register()
```

and later on
`driver.unregister()`

But I don't see how unregister can be used reliable in this static context, since in fact you would either have the static registration (but never unregistration) or control it your self but not have it on a static way...

Reply:
There is no need to register the driver with DriverManager for every instance of SQLServerDriver, hence this method is static.

Reply:
Yes but making this nonstatic would be much cleaner, the driver static initializer can still be the only one registering a driver per default.

Reply:
Hi @laeubi 

Registered driver must be static as there must always be 1 instance of driver registered with DriverManager in a JVM run. It not only prevents resource leaks, but also ensures registeration and deregistration of driver with DriverManager is always synchronized pointing to single instance.

Making it non-static is not possible as Java does not allow accessing non-static objects from static blocks/methods.

Reply:
@cheenamalhotra the register/unregister methods could be instance methods and check against a static variable if there is already a registration. At the moment, everyone can unregister the driver even if he does not gets hold of the actual instance, and you can't get sure that you (accidentally) deregister an instance that was registered by someone else.
I will provide an example...

Beside that, according to the [documentation](https://docs.oracle.com/javase/8/docs/api/java/sql/DriverManager.html) Drivermanager.register SHOULD - so it is optional, you can better use the SPI approach (what is, by the way JDBC 4.0 drivers MUST include such a file!)

> JDBC 4.0 Drivers must include the file META-INF/services/java.sql.Driver

and with the SPI you don't need to call registerDriver anymore since the JVM will find it automatically and there is no risk of a resource leak.

Reply:
Hi @laeubi 

We already support SPI, and registering driver is no longer _required_. Yes, it is taken care by JVM.

Although I was of the view, starting activator also registers the driver, but seems it only instantiates it. I added a commit that explicitly calls `register()` method of `SQLServerDriver` class to attempt registering which will be a safe call and useful if driver has been deregistered and needs to be registered again, and thus does not lead to any classloader conflict issues.